node/node_syncing.go

```diff
@@ -200,18 +200,30 @@ func (node *Node) CalculateResponse(request *downloader_pb.DownloaderRequest) (*
 	response := &downloader_pb.DownloaderResponse{}
 	switch request.Type {
 	case downloader_pb.DownloaderRequest_HEADER:
-		var startHeaderHash []byte
 		if request.BlockHash == nil {
-			tmp := node.Blockchain().Genesis().Hash()
-			startHeaderHash = tmp[:]
-		} else {
-			startHeaderHash = request.BlockHash
+			return response, fmt.Errorf("[SYNC] GetBlockHashes Request BlockHash is NIL")
+		}
+		if request.Size == 0 || request.Size > syncing.BatchSize {
+			return response, fmt.Errorf("[SYNC] GetBlockHashes Request contains invalid Size %v", request.Size)
+		}
+		size := uint64(request.Size)
+		var startHashHeader common.Hash
+		copy(startHashHeader[:], request.BlockHash[:])
+		startBlock := node.Blockchain().GetBlockByHash(startHashHeader)
+		if startBlock == nil {
+			return response, fmt.Errorf("[SYNC] GetBlockHashes Request cannot find startHash %v", startHashHeader)
 		}
-		for block := node.Blockchain().CurrentBlock(); block != nil; block = node.Blockchain().GetBlockByHash(block.Header().ParentHash) {
+		startHeight := startBlock.NumberU64()
+		endHeight := node.Blockchain().CurrentBlock().NumberU64()
+		if startHeight >= endHeight {
+			return response, fmt.Errorf("[SYNC] GetBlockHashes Request failed. I am not higher than requested node, my Height %v, request node Height %v", endHeight, startHeight)
+		}
+		if endHeight-startHeight > size {
+			endHeight = startHeight + size
+		}
+		for blockNum := startHeight; blockNum <= endHeight; blockNum++ {
+			block := node.Blockchain().GetBlockByNumber(blockNum)
```

Comment:
Check `block == nil` here.  In fact, we do not have to loop to `endHeight`, but simply start at `startHeight` and keep going until `len(response.Payload) >= size` or `block == nil` (which indicates that the chain has ended at `blockNum-1`).

Reply:
good point!